### PR TITLE
Document admin bootstrapping and recovery

### LINKED ISSUE #30011: Document admin bootstrapping and recovery
The documentation should cover:
* The [`kc.sh|bat bootstrap-admin user`](https://github.com/keycloak/keycloak/issues/29324) and [`kc.sh|bat bootstrap-admin service`](https://github.com/keycloak/keycloak/issues/30002) commands.
* The [startup env vars](https://github.com/keycloak/keycloak/issues/30003).
* [Operator](https://github.com/keycloak/keycloak/issues/30004).
* That admin recovery can be performed through admin bootstrapping.
  * Example of how realms with increased security (OTP requirement, passwordles, ...) can use the admin service account to recover access through Admin CLI. 
* Release notes and migration notes for startup env vars.

### PR DESCRIPTION
**Pull request by Pepo48**
Closes: #30011

Signed-off-by: Peter Zaoral <pzaoral@redhat.com>

### CODE REVIEW

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [inline docs/guides/server/bootstrap-admin-recovery.adoc:7]: As I mentioned during the meeting, initially I tried: 
```suggestion
summary="Learn how to bootstrap and recover admin account." 
includedOptions="bootstrap-admin-*">
```
but the Relevant options table ended up to be empty. Any idea why?

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:11]: This paragraph is good but possibly should be moved to release notes. IMHO guides should not reminiscence on some previous state of things.

Instead, I'd suggest some introduction that would explain the user the concept of temporary admin user (that they should create a more permanent and secure user etc.).

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:25]: We do not automatically expire the user at this phase, see https://github.com/keycloak/keycloak/issues/30008#issuecomment-2151956072.
```suggestion
This command creates a temporary admin user with the username `tmpadm` and the password `tmpadmpass`.
```

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:53]: We could probably use kcadm.sh|bat in these examples right away.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:9]: Bootstrapping is not just for disaster recovery scenarios.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:45]:
```suggestion
Passwordless, OTP or other advanced authentication methods can be enforced for the realm with lost admin access. In such case, the admin service account needs to be created to recover lost admin access to such realm. After the service account is created, authentication against the {project_name} instance is required to perform all the necessary operations:
```

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:1]: It'd be also good to add a section for bootstrapping at server startup. Explaining the difference from the dedicated command (that on startup master realm must on exist, making the dedicated command the only suitable option for recovery).

We should also add some note that the dedicated command requires the server to be stopped and that it requires the same args as start (db options etc.).

**Review by vmuzikar — COMMENTED** [review]
@Pepo48 Thank you for the draft. I think it's a good start.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc:25]: @vmuzikar you probably reviewed an older revision - at the time you posted the review the text didn't contain any reference regarding the expiration. I immediately cleaned it up after our discussion at the meeting.

Thanks for pointing it out in the fist place.

Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: This was a leftover I forgot about. Adressed, thanks!

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Adressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Adressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Adressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [inline docs/guides/server/bootstrap-admin-recovery.adoc:27]: > It'd be also good to add a section for bootstrapping at server startup. Explaining the difference from the dedicated command (that on startup master realm must on exist, making the dedicated command the only suitable option for recovery).

I tried to address this one via a note.
@shawkins wdyt?

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed. Thanks @shawkins.

**Comment by Pepo48** [comment]
I tried to incorporate every suggestion mentioned above. For now the changes are in a separate commit in order to make it easier to see the differences. I will do the squash later.

cc: @shawkins

**Review by shawkins — DISMISSED** [review]
> I tried to incorporate every suggestion mentioned above. For now the changes are in a separate commit in order to make it easier to see the differences. I will do the squash later.
> 
> cc: @shawkins
> 
> Edit: I also tried to adjust the [punctuation, articles and wording](https://github.com/keycloak/keycloak/compare/4f47c6e66396f099aa240643de797a26ddb16b6d..ff2d3d9d81f8bb7790cb5bbe265209031b032df2).

Thanks @Pepo48 it looks good. Just a conflict to clear up and it should be set.

**Comment by Pepo48** [comment]
> Just a conflict to clear up and it should be set.

@shawkins thanks for the review. The conflict happened in the release notes - I merged both of our inputs together into one section.

**Review by shawkins — DISMISSED** [review]
> > Just a conflict to clear up and it should be set.
> 
> @shawkins thanks for the review. The conflict happened in the release notes - I merged both of our inputs together into one section.

Looks good. Do you want to wait for @vmuzikar to also review?

**Comment by Pepo48** [comment]
@shawkins last week I was told that the whole bootstrap admin and recovery initiative is not considered as a blocker for the upcoming major release. That being said, we don't have to rush it, personally I just wanted to have the bits and pieces ready when they return. (to be able to react quickly then, if for whatever reason necessary)

tl;dr I'm fine with both options 😊

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (ahus1) [inline docs/guides/server/bootstrap-admin-recovery.adoc:27]: IMHO the `db` options would not only be "strongly recommended", but "required" as it would otherwise not write to the correct database and it will not work. 

At the same time I am not sure if all of the options I use on the start command will also work on the boostrap commands (like: setting a HTTP port).

**Comment by ahus1** [comment]
@Pepo48 - I pushed one commit to make the external link check pass. I also found that anchors in the IDs don't work, so I updated the Freemarker makros to support it. It is already part of the scaling guide, but not yet in main. 

If you consider this documentation ready-to-be-merged, please ask Andrew Munro for a review from the technical writer's perspective to make sure the wording aligns with all the other docs.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (shawkins) [reply docs/guides/server/bootstrap-admin-recovery.adoc:27]: This is a somewhat confusing area in general for us #29200 - we do some configuration exclusions for the CLI, but those same options will still work via the other config sources (the operator effectively uses this for export because almost everything is set via an env variable). Only a handful of options have hard-coded settings when running a non-server command.

Comment (shawkins) [reply docs/guides/server/bootstrap-admin-recovery.adoc:27]: I tried to clarify the cases with https://github.com/keycloak/keycloak/issues/29200#issuecomment-2233350379 - as there are potentially 3 different ways users could be invoking non-server commands which have some nuance that we trying to address here quickly with this recommendation.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [reply docs/guides/server/bootstrap-admin-recovery.adoc:1]: > We should also add some note that the dedicated command requires the server to be stopped and that it requires the same args as start (db options etc.).

I believe this is not resolved yet.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:79]: > the client ID is optional and defaults to temp-admin-service

Username also has a default value. Please see the comment above.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@Pepo48 Thank you for the updates. Nice work! Added some comments.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [reply docs/guides/server/bootstrap-admin-recovery.adoc:33]: Note that this applies to service account's secret as well.

Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc:33]: You're correct, I overlooked it for the part in question (as for the service the behaviour should be already documented correctly). I will fix this.

**Comment by vmuzikar** [comment]
@Pepo48 Could you also please address [this comment](https://github.com/keycloak/keycloak/pull/30711#discussion_r1667055511) in this PR since #30711 was merged sooner than this?

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc:49]: What I meant by the statement is that you can create the service first and once you open the admin console, you will be still prompted to create the default admin user account. The formulation is weird mainly because at the time of the writing I considered the default admin and the temporary admin to be some sort of different (thus even the NOTE section you rewrote earlier). Only the UI work made me realize that the default admin is also considered to be temporary. Thanks for noticing this, this needs to be rewritten as well.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc:83]: Addressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Fixed.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed by slightly changing the section structure for the dedicated command section of the guide.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc:99]: I believe this was already partially addressed (I just additionally mentioned the requirement for the server to be stopped) - the question is about the appropriate wording here (see Alex's question and Steven's answer above).

As for the default values, I added a new section.

**Comment by Pepo48** [comment]
> @Pepo48 Could you also please address [this comment](https://github.com/keycloak/keycloak/pull/30711#discussion_r1667055511) in this PR since #30711 was merged sooner than this?

Addressed by adding a new section with the link to the operator's advanced configuration.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:37]:
```suggestion
If no other parameters are specified and/or no corresponding environment variables are set, the user is prompted to enter the required information. The username value can be omitted to use the default values. For more information, see the <<Default values>> and <<Environment variables>> sections below.
```

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:53]:
```suggestion
Similarly, if no corresponding environment variables or additional parameters are set, the user will be prompted to enter the required information. The client ID value can be omitted to use the default values. For more information, see the <<Default values>> and <<Environment variables>> sections below.
```

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [inline docs/guides/server/bootstrap-admin-recovery.adoc:61]: I'd do it the other way around to keep it consistent with other guides – instead of add links to server guide pointing to the operator guide, I'd suggest adding a link to the operator guide pointing to server guide.

**Review by vmuzikar — COMMENTED** [review]
@Pepo48 Thanks for another update. Added some last thoughts, mostly nitpics.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (mhajas) [inline docs/guides/server/bootstrap-admin-recovery.adoc:11]: Could you elaborate more on what temporary means here? Is the access available until the Keycloak node that was started with the parameters is alive?

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (mhajas) [inline docs/guides/server/bootstrap-admin-recovery.adoc:27]: How does this work with Keycloak cluster? Do all nodes need to be stopped before this procedure is executed?

Comment (vmuzikar) [reply docs/guides/server/bootstrap-admin-recovery.adoc:27]: Yes, all nodes need to be stopped:
https://github.com/keycloak/keycloak/blob/1e14b310669126d2dba3d20c902054f3f253ca61/docs/guides/server/bootstrap-admin-recovery.adoc?plain=1#L27

Maybe we could more emphasis on it, make it a warn box or something.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (vmuzikar) [reply docs/guides/server/bootstrap-admin-recovery.adoc:11]: By "temporary" we mean "considered to be temporary by the user". Basically it's just semantical for now and only highlighted as a [warning](https://github.com/keycloak/keycloak/pull/31387) in the UI and logs. In the future, the accounts will be [expired](https://github.com/keycloak/keycloak/issues/30008#issuecomment-2151956072).

Comment (mhajas) [reply docs/guides/server/bootstrap-admin-recovery.adoc:11]: I see. Should we mention in the docs they need to delete the temporary users manually?

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (mhajas) [reply docs/guides/server/bootstrap-admin-recovery.adoc:27]: For example, changing it to "all {project_name} nodes"?

Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed as per suggestion.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (Pepo48) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Addressed.

File: docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc
Comment (Pepo48) [reply docs/documentation/upgrading/topics/changes/changes-26_0_0.adoc:97]: I left the paragraph there and added a new slightly different one to the release notes.

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (mhajas) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Thank you @Pepo48

File: docs/guides/server/bootstrap-admin-recovery.adoc
Comment (mhajas) [reply docs/guides/server/bootstrap-admin-recovery.adoc]: Thank you @Pepo48

**Review by vmuzikar — COMMENTED** [review]
@Pepo48 Thanks for the latest changes. Overall LGTM, added some last minor comments.

@ahus1 @shawkins @mhajas Do you want to re-review as well?

I'd like to merge this rather sooner than later. Not sure we need to wait for a review from @andymunro (as he's quite busy). We could incorporate his review changes as a follow up. WDYT?

File: docs/guides/operator/advanced-configuration.adoc
Comment (Pepo48) [reply docs/guides/operator/advanced-configuration.adoc:317]: what's the general approach here? Should we use a hard-coded links or should we alway try to use variables and link anchors when cross-referencing.

Am I actually allowed to define and use new [document attributes](https://github.com/keycloak/keycloak/pull/31039/files#diff-8fc1f621a47f49335dec8370bcbd8d9e986d76f49442d412c191bc0754a9b75eR64-R65) for this specific case?

Comment (vmuzikar) [reply docs/guides/operator/advanced-configuration.adoc:317]: I'm not sure. Possibly @ahus1 has some input?

Comment (Pepo48) [reply docs/guides/operator/advanced-configuration.adoc:317]: I guess for the guide itself I can replace it with `<@links.server id="bootstrap-admin-recovery"/>`. But what about the [release notes](https://github.com/keycloak/keycloak/pull/31039/files#diff-0519b2567941e80155425829531006542bea0fbf514b7035e421f3f7a7c936e1R77) then?

**Comment by mhajas** [comment]
@vmuzikar, I went through the changes quickly and it looks good. Still, I am not 100% confident to approve so I will leave it for the others to add final approvals. If you would like me to have a proper look let me know.

File: docs/guides/operator/advanced-configuration.adoc
Comment (vmuzikar) [reply docs/guides/operator/advanced-configuration.adoc:317]: Release notes seem to work fine for my local build.

Comment (Pepo48) [reply docs/guides/operator/advanced-configuration.adoc:317]: For me as well. The question is whether it's fine to use it that way.

Comment (vmuzikar) [reply docs/guides/operator/advanced-configuration.adoc:317]: > For me as well. The question is whether it's fine to use it that way.

For release note and upgrading guide, sure from my PoV. For Operator guide, I'm not sure – locally it doesn't work.

Comment (Pepo48) [reply docs/guides/operator/advanced-configuration.adoc:317]: @vmuzikar are the rest of the links in the operator's advanced configuration guide rendered correctly for your local build?

Comment (vmuzikar) [reply docs/guides/operator/advanced-configuration.adoc]: Yes, other links render fine for me.

File: docs/guides/operator/advanced-configuration.adoc
Comment (vmuzikar) [reply docs/guides/operator/advanced-configuration.adoc:1]: #31646 was now merged. @Pepo48 Can you please rebase?

Comment (Pepo48) [reply docs/guides/operator/advanced-configuration.adoc:1]: @vmuzikar rebased.

**Review by vmuzikar — APPROVED** [review]
@Pepo48 After the latest changes, LGTM! Thank you!

@shawkins @ahus1 Do you want to review as well?

**Review by shawkins — APPROVED** [review]
LGTM, thank you @Pepo48

**Pull request merged**
